File: Cauly3.1/MediationSample/CaulyObjectiveCMediationSample/CaulyCustomEvent/CaulyEventNative.m
//
//  CaulyEventNative.m
//  CaulyIOSMediationSample
//
//  Created by FSN on 2023/03/15.
//

#import "CaulyEventNative.h"


@implementation CaulyEventNative

/// Used to store the ad's images. In order to implement the GADMediationNativeAd protocol, we use
/// this class to return the images property.
NSArray<GADNativeAdImage *> *_images;

/// Used to store the ad's icon. In order to implement the GADMediationNativeAd protocol, we use
/// this class to return the icon property.
GADNativeAdImage *_icon;

/// Used to store the ad's ad choices view. In order to implement the GADMediationNativeAd protocol,
/// we use this class to return the adChoicesView property.
UIView *_adChoicesView;

UIImageView *_mediaView;

UIImage *imageImg;

#pragma mark - Native Ad Mapping
- (nullable NSString *)headline {
    return [nativeAdItem objectForKey:@"title"];
}

- (nullable NSArray<GADNativeAdImage *> *)images {
    return _images;
}

- (nullable NSString *)body {
    return [nativeAdItem objectForKey:@"description"];
}

- (nullable GADNativeAdImage *)icon {
    NSLog(@"%@", _icon.image);
    return _icon;
}

- (nullable NSString *)callToAction {
    return nil;
}

- (nullable NSDecimalNumber *)starRating {
  return nil;
}

- (nullable NSString *)store {
  return nil;
}

- (nullable NSString *)price {
  return nil;
}

- (nullable NSString *)advertiser {
    return [nativeAdItem objectForKey:@"subtitle"];
}

- (nullable NSDictionary<NSString *, id> *)extraAssets {
    return nil;
}

- (nullable UIView *)adChoicesView {
  return _adChoicesView;
}

- (nullable UIView *)mediaView {
    return _mediaView;
}

- (BOOL)hasVideoContent {
  return self.mediaView != nil;
}

#pragma mark - Native Ad Request
- (void)loadNativeAdForAdConfiguration:(GADMediationNativeAdConfiguration *)adConfiguration completionHandler:(GADMediationNativeLoadCompletionHandler)completionHandler {
    __block GADMediationNativeLoadCompletionHandler originalCompletionHandler =
        [completionHandler copy];

    _loadCompletionHandler = ^id<GADMediationNativeAdEventDelegate>(
        _Nullable id<GADMediationNativeAd> ad, NSError *_Nullable error) {

      id<GADMediationNativeAdEventDelegate> delegate = nil;
      if (originalCompletionHandler) {
        // Call original handler and hold on to its return value.
        delegate = originalCompletionHandler(ad, error);
      }

      // Release reference to handler. Objects retained by the handler will also be released.
      originalCompletionHandler = nil;

      return delegate;
    };
    
    // admob 대시보드 등록 parameter : cauly appCode
    NSString *adUnit = adConfiguration.credentials.settings[@"parameter"];
    
    CaulyAdSetting * adSetting = [CaulyAdSetting globalSetting];
    
    // 카울리 로그 레벨
    [CaulyAdSetting setLogLevel:CaulyLogLevelInfo];
    
    // iTunes App ID
    adSetting.appId = @"0";
    
    // 카울리 앱 코드
    adSetting.appCode = adUnit;
    
    _nativeAd = [[CaulyNativeAd alloc] initWithParentViewController:[adConfiguration topViewController]];
    _nativeAd.delegate = self;
    
    // 카울리 네이티브 광고 요청
    [_nativeAd startNativeAdRequest:2 nativeAdComponentType:CaulyNativeAdComponentType_IconImage imageSize:@"720x480"];
}

#pragma mark - Cauly Native Delegates

// 네이티브 광고 정보 수신 성공
-(void) didReceiveNativeAd:(CaulyNativeAd *)nativeAd isChargeableAd:(BOOL)isChargeableAd {
    NSLog(@"Cauly didReceiveNativeAd");
    CaulyNativeAdItem *caulyNativeAd = [nativeAd nativeAdItemAt:0];
    caulyNativeAdItem = [nativeAd nativeAdItemAt:0];
    
    NSArray* allList = [nativeAd nativeAdItemList];
    
    for (CaulyNativeAdItem *adItem in allList) {
        // 수신된 모든 네이티브 광고(JSON) 로그 출력
        NSLog(@"for nativeAdJSONString : %@", adItem.nativeAdJSONString);
    }
    
    NSError *error;
    nativeAdItem = [NSJSONSerialization JSONObjectWithData:[caulyNativeAd.nativeAdJSONString dataUsingEncoding:NSUTF8StringEncoding] options:kNilOptions error:&error];
    
    _nativeAd = nativeAd;
    
    [self performSelectorInBackground:@selector(urlToImage) withObject:nil];
    
}

- (void)didFailToReceiveNativeAd:(CaulyNativeAd *)nativeAd errorCode:(int)errorCode errorMsg:(NSString *)errorMsg {
    NSLog(@"Cauly didFailToReceiveNativeAd errorCode: %d errorMsg: %@", errorCode, errorMsg);
    NSError *error = [[NSError alloc] initWithDomain:@"kr.co.cauly.sdk.ios.mediation.sample" code:errorCode userInfo:@{NSLocalizedDescriptionKey:errorMsg}];
    _adEventDelegate = _loadCompletionHandler(nil, error);
}

// 이미지 url 변환
- (void)urlToImage {
    NSURL *iconURL = [NSURL URLWithString:[nativeAdItem objectForKey:@"icon"]];
    NSData *iconData = [NSData dataWithContentsOfURL:iconURL];
    UIImage *iconImg = [[UIImage alloc] initWithData:iconData];
    _icon = [[GADNativeAdImage alloc] initWithImage:iconImg];
    
    NSURL *imageURL = [NSURL URLWithString:[nativeAdItem objectForKey:@"image"]];
    NSData *imageData = [NSData dataWithContentsOfURL:imageURL];
    imageImg = [[UIImage alloc] initWithData:imageData];

    [self performSelectorOnMainThread:@selector(setMediaViewImage) withObject:nil waitUntilDone:0];
}

// 변환된 이미지를 MediaView 에 표시
- (void)setMediaViewImage {
    _mediaView = [[UIImageView alloc] initWithImage:imageImg];
    _adEventDelegate = _loadCompletionHandler(self, nil);
}

#pragma mark - GADMediatedUnifiedNativeAd

- (void)didRecordClickOnAssetWithName:(GADNativeAssetIdentifier)assetName view:(UIView *)view viewController:(UIViewController *)viewController {
    NSLog(@"didRecordClickOnAssetWithName");
    [_nativeAd click:caulyNativeAdItem];
}

@end

File: Cauly3.1/MediationSample/CaulySwiftMediationSample/CaulyCustomEvent/CaulyEventNative.swift
//
//  CaulyEventNative.swift
//  CaulySwiftMediationSample
//
//  Created by FSN on 2023/07/24.
//

import Foundation
import GoogleMobileAds

class CaulyEventNative: NSObject, GADMediationNativeAd, CaulyNativeAdDelegate {
    /// The Sample Ad Network native ad.
    var nativeAd: CaulyNativeAd?
    
    var nativeAdItem: Dictionary<String, Any>?
    
    var caulyNativeAdItem: CaulyNativeAdItem?
    
    var _mediaView: UIView?
    
    /// The ad event delegate to forward ad rendering events to the Google Mobile Ads SDK.
    var delegate: GADMediationNativeAdEventDelegate?
    
    /// Completion handler called after ad load
    var completionHandler: GADMediationNativeLoadCompletionHandler?
    
    func loadNativeAd( for adConfiguration: GADMediationNativeAdConfiguration, completionHandler: @escaping GADMediationNativeLoadCompletionHandler) {
        // admob 등록 parameter(Cauly 발급 키)를 가져옵니다.
        let adUnit = adConfiguration.credentials.settings["parameter"] as? String
        
        // 상세 설정 항목들은 표 참조, 설정되지 않은 항목들은 기본값으로 설정됩니다.
        let caulySetting = CaulyAdSetting.global()
        CaulyAdSetting.setLogLevel(CaulyLogLevelInfo)   // CaulyLog 레벨
        caulySetting?.appId = "0"                       // App Store 에 등록된 App ID 정보
        caulySetting?.appCode = adUnit                  // admob 등록 parameter (Cauly 발급 키)
        
        nativeAd = CaulyNativeAd.init(parentViewController: adConfiguration.topViewController)
        nativeAd?.delegate = self;
        self.completionHandler = completionHandler
        
        // 카울리 네이티브 광고 요청
        nativeAd?.startRequest(2, nativeAdComponentType: CaulyNativeAdComponentType_IconImage, imageSize: "720x480")
    }
    
    // MARK: - GADMediationNative
    required override init() {
        super.init()
    }
    
    var headline: String? {
        return nativeAdItem?["title"] as? String
    }
    
    var images: [GADNativeAdImage]?
    
    var body: String? {
        return nativeAdItem?["description"] as? String
    }
    
    var icon: GADNativeAdImage?
    
    var callToAction: String? {
        return ""
    }
    
    var starRating: NSDecimalNumber? {
        return nil
    }
    
    var store: String? {
        return ""
    }
    
    var price: String? {
        return ""
    }
    
    var advertiser: String? {
        return nativeAdItem?["subtitle"] as? String
    }
    
    var extraAssets: [String : Any]? {
        return nil
    }
    
    var adChoicesView: UIView?
    
    var mediaView: UIView? {
        return _mediaView
    }
    
    // MARK: - Cauly Native Delegate
    // 네이티브 광고 정보 수신 성공
    func didReceive(_ nativeAd: CaulyNativeAd!, isChargeableAd: Bool) {
        print("Cauly didReceiveNativeAd")
        caulyNativeAdItem = nativeAd.nativeAdItem(at: 0)
        
        do {
            nativeAdItem = try JSONSerialization.jsonObject(with: Data((caulyNativeAdItem?.nativeAdJSONString.utf8)!), options: []) as? Dictionary<String, Any>
        } catch {
            print(error.localizedDescription)
        }
        print(nativeAdItem as Any)
        self.nativeAd = nativeAd
        
        urlToImage()
    }
    
    func didFail(toReceive nativeAd: CaulyNativeAd!, errorCode: Int32, errorMsg: String!) {
        print("Cauly didFailToReceiveNativeAd errorCode: \(errorCode), errMsg: \(errorMsg ?? "No Message")")
        
        let error = NSError(domain: "kr.co.cauly.sdk.ios.mediation.sample", code: Int(errorCode), userInfo: ["description" : errorMsg as Any])
        
        if let handler = completionHandler {
            delegate = handler(nil, error)
        }
    }
    
    func urlToImage() {
        let iconUrl = URL(string: nativeAdItem?["icon"] as! String)
        let mediaUrl = URL(string: nativeAdItem?["image"] as! String)
        
        DispatchQueue.global().async { [weak self] in
            if let iconData = try? Data(contentsOf: iconUrl!), let mediaData = try? Data(contentsOf: mediaUrl!) {
                if let iconImage = UIImage(data: iconData), let mediaImage = UIImage(data: mediaData) {
                    DispatchQueue.main.async {
                        self?.icon = GADNativeAdImage(image: iconImage)
                        self?._mediaView = UIImageView(image: mediaImage)
                        self?.images = [GADNativeAdImage(image: mediaImage)]
                        
                        if let handler = self?.completionHandler {
                            self?.delegate = handler(self, nil)
                        }
                    }
                }
            }
        }
    }
}
